[DataPipe] Change DataPipe display name in profiler

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100042

This PR improves how DataPipe operations are displayed in performance profiler for better diagnostics.

Script:
```python
from torchdata.datapipes.iter import IterableWrapper
from torchdata.dataloader2 import DataLoader2, MultiProcessingReadingService

ls = range(16)
dp = IterableWrapper(ls).map(fn_2).map(fn_3).map(fn_4)

rs = MultiProcessingReadingService(num_workers=0, main_prefetch_cnt=0, worker_prefetch_cnt=0)
dl2 = DataLoader2(dp, reading_service=rs)

with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU]) as prof:
    for x in dl2:
        pass
```


Output before:
```
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                               Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
             enumerate(DataPipe)#MapperIterDataPipe        76.37%       1.419ms       213.08%       3.959ms      80.796us            49  
    enumerate(DataPipe)#IterableWrapperIterDataPipe        12.70%     236.000us        12.70%     236.000us      13.882us            17  
...
```

Output after:
```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Mapper(datapipe=Mapper, fn=fn_4, input_col=None, out...        29.79%     645.000us        99.17%       2.147ms     126.294us            17  
Mapper(datapipe=IterableWrapper, fn=fn_2, input_col=...        29.24%     633.000us        42.96%     930.000us      54.706us            17  
Mapper(datapipe=Mapper, fn=fn_3, input_col=None, out...        24.76%     536.000us        68.59%       1.485ms      87.353us            17  
IterableWrapper(deepcopy=True, iterable=range(0, 16)...        10.58%     229.000us        10.58%     229.000us      13.471us            17  
...
```

NivekT committed:
[DataPipe] Change DataPipe display name in profiler

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/_hook_iterator.py:
```diff
@@ -138,7 +143,7 @@ def __next__(self):
             # See: https://github.com/pytorch/data/issues/284
             if self._profiler_enabled:
-                with profiler_record_fn_context():
+                with profiler_record_fn_context(self.source_dp):
```
Is `self.source_dp` the `DataPipe` that generates this iterator or the one prior to this `DataPipe`?

ejguan replied:
Okay, I think it's the `DataPipe` that generates this iterator. Can we change the name from `source_dp` to `datapipe`?

ejguan reviewed (approved):
The change looks good to me. Can you pls share a screenshot of a profiling result in the PR summary?

NivekT committed:
Update on "[DataPipe] Change DataPipe display name in profiler"


Script:
```python
from torchdata.datapipes.iter import IterableWrapper
from torchdata.dataloader2 import DataLoader2, MultiProcessingReadingService

ls = range(16)
dp = IterableWrapper(ls).map(fn_2).map(fn_3).map(fn_4)

rs = MultiProcessingReadingService(num_workers=0, main_prefetch_cnt=0, worker_prefetch_cnt=0)
dl2 = DataLoader2(dp, reading_service=rs)

with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU]) as prof:
    for x in dl2:
        pass
```


Output before:
```
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                               Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
             enumerate(DataPipe)#MapperIterDataPipe        76.37%       1.419ms       213.08%       3.959ms      80.796us            49  
    enumerate(DataPipe)#IterableWrapperIterDataPipe        12.70%     236.000us        12.70%     236.000us      13.882us            17  
...
```

Output after:
```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Mapper(datapipe=Mapper, fn=fn_4, input_col=None, out...        29.79%     645.000us        99.17%       2.147ms     126.294us            17  
Mapper(datapipe=IterableWrapper, fn=fn_2, input_col=...        29.24%     633.000us        42.96%     930.000us      54.706us            17  
Mapper(datapipe=Mapper, fn=fn_3, input_col=None, out...        24.76%     536.000us        68.59%       1.485ms      87.353us            17  
IterableWrapper(deepcopy=True, iterable=range(0, 16)...        10.58%     229.000us        10.58%     229.000us      13.471us            17  
...
```

[ghstack-poisoned]

NivekT commented on torch/utils/data/datapipes/_hook_iterator.py:
```diff
@@ -99,27 +106,30 @@ def _set_datapipe_valid_iterator_id(datapipe):
-    def profiler_record_fn_context():
-        return torch.autograd.profiler.record_function(profile_name)
+
+    def profiler_record_fn_context(datapipe):
+        if not hasattr(datapipe, "_profile_name"):
+            datapipe._profile_name = _generate_iterdatapipe_msg(datapipe, simplify_dp_name=True)
+        return torch.autograd.profiler.record_function(datapipe._profile_name)
```
`_profile_name` is generated once for each instance of `IterDataPipe` the first time `__iter__` is called on the instance.

It is *not* shared among instances of the same class. See PR description for an example.

NivekT commented on torch/utils/data/datapipes/_hook_iterator.py:
```diff
@@ -99,27 +106,30 @@ def _set_datapipe_valid_iterator_id(datapipe):
         method commonly returns `self` but not necessarily.
         """
-        def __init__(self, iterator, source_dp, iterator_id, has_next_method):
+        def __init__(self, iterator, datapipe, iterator_id, has_next_method):
```
@ejguan I changed the name from `source_dp` to `datapipe` per your comment.

NivekT reviewed (commented):
@ejguan I made some changes and updated the PR description to show an example of before/after.

Note, if a DataPipe is invoked twice with the same exact arguments (i.e. `dp.map(fn1).map(fn1)`, the two instances of `map` will have the same `_profile_name` and their entries will be combined within the profiler statistics table.  I think that should be rare and if that happens, it is fine to combine them?

One solution is to add memory address as part of `_profile_name`. Let me know what you think.

ejguan commented:
> Note, if a DataPipe is invoked twice with the same exact arguments (i.e. `dp.map(fn1).map(fn1)`, the two instances of `map` will have the same `_profile_name` and their entries will be combined within the profiler statistics table. I think that should be rare and if that happens, it is fine to combine them?

I think it's fine for now unless there is a specific user request.

ejguan reviewed (approved):
LGTM. Thank you

NivekT committed:
Update on "[DataPipe] Change DataPipe display name in profiler"


Script:
```python
from torchdata.datapipes.iter import IterableWrapper
from torchdata.dataloader2 import DataLoader2, MultiProcessingReadingService

ls = range(16)
dp = IterableWrapper(ls).map(fn_2).map(fn_3).map(fn_4)

rs = MultiProcessingReadingService(num_workers=0, main_prefetch_cnt=0, worker_prefetch_cnt=0)
dl2 = DataLoader2(dp, reading_service=rs)

with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU]) as prof:
    for x in dl2:
        pass
```


Output before:
```
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                               Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
             enumerate(DataPipe)#MapperIterDataPipe        76.37%       1.419ms       213.08%       3.959ms      80.796us            49  
    enumerate(DataPipe)#IterableWrapperIterDataPipe        12.70%     236.000us        12.70%     236.000us      13.882us            17  
...
```

Output after:
```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Mapper(datapipe=Mapper, fn=fn_4, input_col=None, out...        29.79%     645.000us        99.17%       2.147ms     126.294us            17  
Mapper(datapipe=IterableWrapper, fn=fn_2, input_col=...        29.24%     633.000us        42.96%     930.000us      54.706us            17  
Mapper(datapipe=Mapper, fn=fn_3, input_col=None, out...        24.76%     536.000us        68.59%       1.485ms      87.353us            17  
IterableWrapper(deepcopy=True, iterable=range(0, 16)...        10.58%     229.000us        10.58%     229.000us      13.471us            17  
...
```

[ghstack-poisoned]

NivekT commented:
@pytorchbot merge -r

NivekT committed:
Update on "[DataPipe] Change DataPipe display name in profiler"


Script:
```python
from torchdata.datapipes.iter import IterableWrapper
from torchdata.dataloader2 import DataLoader2, MultiProcessingReadingService

ls = range(16)
dp = IterableWrapper(ls).map(fn_2).map(fn_3).map(fn_4)

rs = MultiProcessingReadingService(num_workers=0, main_prefetch_cnt=0, worker_prefetch_cnt=0)
dl2 = DataLoader2(dp, reading_service=rs)

with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU]) as prof:
    for x in dl2:
        pass
```


Output before:
```
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                               Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
             enumerate(DataPipe)#MapperIterDataPipe        76.37%       1.419ms       213.08%       3.959ms      80.796us            49  
    enumerate(DataPipe)#IterableWrapperIterDataPipe        12.70%     236.000us        12.70%     236.000us      13.882us            17  
...
```

Output after:
```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Mapper(datapipe=Mapper, fn=fn_4, input_col=None, out...        29.79%     645.000us        99.17%       2.147ms     126.294us            17  
Mapper(datapipe=IterableWrapper, fn=fn_2, input_col=...        29.24%     633.000us        42.96%     930.000us      54.706us            17  
Mapper(datapipe=Mapper, fn=fn_3, input_col=None, out...        24.76%     536.000us        68.59%       1.485ms      87.353us            17  
IterableWrapper(deepcopy=True, iterable=range(0, 16)...        10.58%     229.000us        10.58%     229.000us      13.471us            17  
...
```

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Change DataPipe display name in profiler"


Script:
```python
from torchdata.datapipes.iter import IterableWrapper
from torchdata.dataloader2 import DataLoader2, MultiProcessingReadingService

ls = range(16)
dp = IterableWrapper(ls).map(fn_2).map(fn_3).map(fn_4)

rs = MultiProcessingReadingService(num_workers=0, main_prefetch_cnt=0, worker_prefetch_cnt=0)
dl2 = DataLoader2(dp, reading_service=rs)

with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU]) as prof:
    for x in dl2:
        pass
```


Output before:
```
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                               Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
             enumerate(DataPipe)#MapperIterDataPipe        76.37%       1.419ms       213.08%       3.959ms      80.796us            49  
    enumerate(DataPipe)#IterableWrapperIterDataPipe        12.70%     236.000us        12.70%     236.000us      13.882us            17  
...
```

Output after:
```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Mapper(datapipe=Mapper, fn=fn_4, input_col=None, out...        29.79%     645.000us        99.17%       2.147ms     126.294us            17  
Mapper(datapipe=IterableWrapper, fn=fn_2, input_col=...        29.24%     633.000us        42.96%     930.000us      54.706us            17  
Mapper(datapipe=Mapper, fn=fn_3, input_col=None, out...        24.76%     536.000us        68.59%       1.485ms      87.353us            17  
IterableWrapper(deepcopy=True, iterable=range(0, 16)...        10.58%     229.000us        10.58%     229.000us      13.471us            17  
...
```

[ghstack-poisoned]

NivekT commented:
@pytorchbot merge -r

NivekT committed:
Update on "[DataPipe] Change DataPipe display name in profiler"


Script:
```python
from torchdata.datapipes.iter import IterableWrapper
from torchdata.dataloader2 import DataLoader2, MultiProcessingReadingService

ls = range(16)
dp = IterableWrapper(ls).map(fn_2).map(fn_3).map(fn_4)

rs = MultiProcessingReadingService(num_workers=0, main_prefetch_cnt=0, worker_prefetch_cnt=0)
dl2 = DataLoader2(dp, reading_service=rs)

with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU]) as prof:
    for x in dl2:
        pass
```


Output before:
```
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                               Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
             enumerate(DataPipe)#MapperIterDataPipe        76.37%       1.419ms       213.08%       3.959ms      80.796us            49  
    enumerate(DataPipe)#IterableWrapperIterDataPipe        12.70%     236.000us        12.70%     236.000us      13.882us            17  
...
```

Output after:
```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Mapper(datapipe=Mapper, fn=fn_4, input_col=None, out...        29.79%     645.000us        99.17%       2.147ms     126.294us            17  
Mapper(datapipe=IterableWrapper, fn=fn_2, input_col=...        29.24%     633.000us        42.96%     930.000us      54.706us            17  
Mapper(datapipe=Mapper, fn=fn_3, input_col=None, out...        24.76%     536.000us        68.59%       1.485ms      87.353us            17  
IterableWrapper(deepcopy=True, iterable=range(0, 16)...        10.58%     229.000us        10.58%     229.000us      13.471us            17  
...
```

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Change DataPipe display name in profiler"


Script:
```python
from torchdata.datapipes.iter import IterableWrapper
from torchdata.dataloader2 import DataLoader2, MultiProcessingReadingService

ls = range(16)
dp = IterableWrapper(ls).map(fn_2).map(fn_3).map(fn_4)

rs = MultiProcessingReadingService(num_workers=0, main_prefetch_cnt=0, worker_prefetch_cnt=0)
dl2 = DataLoader2(dp, reading_service=rs)

with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU]) as prof:
    for x in dl2:
        pass
```


Output before:
```
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                               Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
             enumerate(DataPipe)#MapperIterDataPipe        76.37%       1.419ms       213.08%       3.959ms      80.796us            49  
    enumerate(DataPipe)#IterableWrapperIterDataPipe        12.70%     236.000us        12.70%     236.000us      13.882us            17  
...
```

Output after:
```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Mapper(datapipe=Mapper, fn=fn_4, input_col=None, out...        29.79%     645.000us        99.17%       2.147ms     126.294us            17  
Mapper(datapipe=IterableWrapper, fn=fn_2, input_col=...        29.24%     633.000us        42.96%     930.000us      54.706us            17  
Mapper(datapipe=Mapper, fn=fn_3, input_col=None, out...        24.76%     536.000us        68.59%       1.485ms      87.353us            17  
IterableWrapper(deepcopy=True, iterable=range(0, 16)...        10.58%     229.000us        10.58%     229.000us      13.471us            17  
...
```

[ghstack-poisoned]

NivekT commented:
@pytorchbot merge -r

NivekT committed:
Update on "[DataPipe] Change DataPipe display name in profiler"


Script:
```python
from torchdata.datapipes.iter import IterableWrapper
from torchdata.dataloader2 import DataLoader2, MultiProcessingReadingService

ls = range(16)
dp = IterableWrapper(ls).map(fn_2).map(fn_3).map(fn_4)

rs = MultiProcessingReadingService(num_workers=0, main_prefetch_cnt=0, worker_prefetch_cnt=0)
dl2 = DataLoader2(dp, reading_service=rs)

with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU]) as prof:
    for x in dl2:
        pass
```


Output before:
```
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                               Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
             enumerate(DataPipe)#MapperIterDataPipe        76.37%       1.419ms       213.08%       3.959ms      80.796us            49  
    enumerate(DataPipe)#IterableWrapperIterDataPipe        12.70%     236.000us        12.70%     236.000us      13.882us            17  
...
```

Output after:
```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Mapper(datapipe=Mapper, fn=fn_4, input_col=None, out...        29.79%     645.000us        99.17%       2.147ms     126.294us            17  
Mapper(datapipe=IterableWrapper, fn=fn_2, input_col=...        29.24%     633.000us        42.96%     930.000us      54.706us            17  
Mapper(datapipe=Mapper, fn=fn_3, input_col=None, out...        24.76%     536.000us        68.59%       1.485ms      87.353us            17  
IterableWrapper(deepcopy=True, iterable=range(0, 16)...        10.58%     229.000us        10.58%     229.000us      13.471us            17  
...
```

[ghstack-poisoned]
